### PR TITLE
Synthesize

### PR DESCRIPTION
**Pull request by tkell**
Needs:

- ~~magic stereo zero-crossing!~~
- ~~the resampling that @bmcfee and @blacker were discussing in Issue #5!~~
- ~~tests~~ and documentation!

### CODE REVIEW

**Comment by tkell** [comment]
The resampling is happening at loadtime:  we're keeping raw_samples for rending 44,100 audio later, and making analysis_samples at 22,050 for use in various features.  

(@bmcfee, does librosa really only resample things in mono?  Or can I not read?)

**Comment by bmcfee** [comment]
> (@bmcfee, does librosa really only resample things in mono? Or can I not read?)

it should work in both stereo and mono.

**Comment by tkell** [comment]
@bmcfee, I was getting an error in valid_audio, which requires things to be in mono, it looked like - which is why I did two resamples.

**Comment by bmcfee** [comment]
> @bmcfee, I was getting an error in valid_audio, which requires things to be in mono, it looked like - which is why I did two resamples.

Dang, you're right.  It works if you're doing resample-on-load, but `librosa.resample` itself is mono-only.

One more thing to add to the todo if i ever get around to writing a resampler library...

File: amen/synthesize.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    # should we also support lists of tuples, zipped by the user?  aiii...
+
+    synthesize(some_generator(time_slices))
+    # assumes a generator that returns tuples of slices and times
+
+    """
+
+    # First we organize our inputs.
+    proper_list = []
+    if type(inputs) == TimingList or type(inputs) == list:
```
Comment (bmcfee) [inline amen/synthesize.py]: don't check type == ; use ``isinstance(inputs, TimingList)``

File: amen/synthesize.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            time_index = time_index + time_slice.duration
+        proper_list = zip(inputs, timings)
+    elif type(inputs) == tuple:
+        proper_list = zip(inputs[0], inputs[1])
+    elif isinstance(inputs, types.GeneratorType):
+        proper_list = inputs
+
+    max_time = 0.0
+    array_shape = (2, 44100 * 60 * 20)
+    sparse_array = np.zeros(array_shape)
```
Comment (bmcfee) [inline amen/synthesize.py]: to make this sparse, do ``scipy.sparse.csr_matrix`` instead of ``np.zeros``

File: amen/synthesize.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    for time_slice, start_time in proper_list:
+        start_time = start_time.delta * 1e-9
+        duration = time_slice.duration.delta * 1e-9
+        if start_time + duration > max_time:
```
Comment (bmcfee) [inline amen/synthesize.py]: throw an exception here

File: amen/time.py [hunk 4]
```diff
@@ -19,6 +20,17 @@ def __repr__(self):
         args = self.time.delta * 1e-9, self.duration.delta * 1e-9
         return '<TimeSlice, start: {0:.2f}, duration: {1:.2f}'.format(*args)
```
Comment (bmcfee) [inline amen/time.py]: missing close >

File: amen/synthesize.py (hunk 3 above)
Comment (tkell) [inline amen/synthesize.py]: done, don't know why GH is not saying so.

**Comment by tkell** [comment]
I think I have the zero-crossing fixes working!  I am going to start thinking about tests for all this, but please put your wisdom / numpy tricks / etc here.

**Comment by tkell** [comment]
...added several tests.  Still need to make a test for the zero-crossing stuff, I think, and do some stereo vs mono cleanup.

**Comment by tkell** [comment]
Well, so much for that cute test.  More to work on this week.  

@bmcfee, it appears that when I write out a file with librosa, I can't read it back in without getting a BackEndError from audioread - but this is crazy, because you have a passing librosa test that does exactly that.

File: .travis_dependencies.sh [hunk 5]
```diff
@@ -41,6 +41,7 @@ if [ ! -d "$src" ]; then
 
         pip install git+https://github.com/bmcfee/samplerate.git
         pip install python-coveralls
+        pip install mock
```
Comment (tkell) [inline .travis_dependencies.sh]: Forgot to take this out, was for the previous version of the to_wav test.

File: amen/examples/reverse.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# encoding: utf=8
+
+"""
+reverse.py : Reverse the beats of a song.
+"""
+
+import sys
+from amen.audio import Audio
+from amen.utils import example_audio_file
+from amen.synthesize import snthesize
```
Comment (tkell) [inline amen/examples/reverse.py]: `synthesize`, not `snthesize`

File: amen/time.py [hunk 7]
```diff
@@ -17,7 +20,40 @@ def __init__(self, time, duration, audio, unit='s'):
...
+        starting_sample, ending_sample = librosa.time_to_samples([start, start + duration], self.audio.sample_rate)
+
+        left_offsets = self._get_offsets(self.audio.raw_samples[0], starting_sample, ending_sample)
+        right_offsets = self._get_offsets(self.audio.raw_samples[1], starting_sample, ending_sample)
+
+        samples = self._offset_samples(starting_sample, ending_sample, left_offsets, right_offsets)
+
+        return samples, left_offsets, right_offsets
+
+    def _get_offsets(self, channel, starting_sample, ending_sample):
```
Comment (tkell) [inline amen/time.py]: I think that this should just do:

`left_offsets, right_offsets = self._get_offsets(starting_sample, ending_sample)`

File: amen/time.py [hunk 8]
```diff
@@ -17,7 +20,40 @@ def __init__(self, time, duration, audio, unit='s'):
...
+
+    def get_samples(self):
+        """
+        Gets the samples corresponding to this TimeSlice from the parent audio object.
+        """
+        start = self.time.delta * 1e-9
+        duration = self.duration.delta * 1e-9
+        starting_sample, ending_sample = librosa.time_to_samples([start, start + duration], self.audio.sample_rate)
+
+        left_offsets = self._get_offsets(self.audio.raw_samples[0], starting_sample, ending_sample)
```
Comment (tkell) [inline amen/time.py]: I think that this should just do:

`left_offsets, right_offsets = self._get_offsets(starting_sample, ending_sample)`

File: amen/synthesize.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        time_index = pd.to_timedelta(0.0, 's')
+        timings = []
+        for time_slice in inputs:
+            timings.append(time_index)
+            time_index = time_index + time_slice.duration
+        proper_list = zip(inputs, timings)
+    elif isinstance(inputs, tuple):
+        proper_list = zip(inputs[0], inputs[1])
+    elif isinstance(inputs, types.GeneratorType):
+        proper_list = inputs
```
Comment (tkell) [inline amen/synthesize.py]: This chunk can become something like `inputs = format_inputs(inputs)`

File: amen/audio.py [hunk 10]
```diff
@@ -49,16 +55,36 @@ def __init__(self, file_path, convert_to_mono=False, sample_rate=22050):
...
         self.file_path = file_path
-        y, sr = librosa.load(file_path, mono=convert_to_mono, sr=sample_rate)
         self.sample_rate = float(sr)
+        self.analysis_sample_rate = float(analysis_sample_rate)
+
         self.raw_samples = y
         self.num_channels = y.ndim
         self.duration = librosa.get_duration(y=y, sr=sr)
+
+        if y.ndim == 2:
```
Comment (bmcfee) [inline amen/audio.py]: librosa 0.4.2 will handle multi-channel resampling properly, so this block will not be necessary.

**Comment by tkell** [comment]
K, I think I've got everything.  I await feedback / fixes / better syntax, etc!

File: amen/audio.py (hunk 10 above)
Comment (tkell) [inline amen/audio.py]: :rocket:
